### PR TITLE
Do not dispose singleton instances

### PR DESCRIPTION
**Pull request by pakrym**
Fixes: https://github.com/aspnet/DependencyInjection/issues/459
@davidfowl @halter73

### CODE REVIEW

File: test/Microsoft.Extensions.DependencyInjection.Tests/ServiceProviderContainerTests.cs [hunk 1]
```diff
@@ -101,9 +101,34 @@ public void CreatingServiceProviderWithUnresolvableTypesThrows(Type serviceType,
...
+        public void DoesNotDisposeSingletonInstances()
+        {
+            var disposable = new Disposable();
+            var serviceCollection = new ServiceCollection();
+            serviceCollection.AddSingleton(disposable);
+
+            var provider = serviceCollection.BuildServiceProvider();
+            provider.GetService<Disposable>();
+
+            (provider as IDisposable)?.Dispose();
```
Comment (halter73) [inline test/Microsoft.Extensions.DependencyInjection.Tests/ServiceProviderContainerTests.cs]: Nit: hard cast

**Review by halter73 — APPROVED** [review]

**Comment by davidfowl** [comment]
We need to see which of our types are going to start leaking because of this. I think the logger factory in hosting will need to be disposed manually.

**Review by davidfowl — APPROVED** [review]

**Comment by khellang** [comment]
> We need to see which of our types are going to start leaking because of this.

Is this really a leak? The application container won't be disposed until the app shuts down anyway, right?

What about all the apps out there that might also "depend" on this behavior? Shouldn't this be classified as a breaking change? :thinking:

**Comment by davidfowl** [comment]
> Is this really a leak? The application container won't be disposed until the app shuts down anyway, right?

Sure in the "default" application. Wait until people start hosting ASP.NET in a shared environment like visual studio. We shouldn't assume disposing the host means the application is done.

> What about all the apps out there that might also "depend" on this behavior? Shouldn't this be classified as a breaking change? 🤔

Yes, it is a breaking change.

/cc @DamianEdwards @Eilon

**Comment by khellang** [comment]
> We shouldn't assume disposing the host means the application is done.

Good point :+1:

**Comment by pakrym** [comment]
@davidfowl what about registering logger provider a as factory returning instance if we know it should be disposed with container?

**Comment by Eilon** [comment]
FYI I marked the bug https://github.com/aspnet/DependencyInjection/issues/459 as breaking-change.

**Comment by khellang** [comment]
You're making breaking changes to 1.1? :confused:

**Comment by pakrym** [comment]
@khellang we are fine with taking this one because fixed behavior is much more useful and in typical application does not break anyone.

**Pull request merged**